plugins/cilium-cni/main.go

```diff
@@ -666,14 +665,14 @@ func cmdDel(args *skel.CmdArgs) error {
 		return err
 	}
 
-	id := endpointid.NewID(endpointid.ContainerIdPrefix, args.ContainerID)
-	if err := c.EndpointDelete(id); err != nil {
-		// EndpointDelete returns an error in the following scenarios:
-		// DeleteEndpointIDInvalid: Invalid delete parameters, no need to retry
-		// DeleteEndpointIDNotFound: No need to retry
-		// DeleteEndpointIDErrors: Errors encountered while deleting,
-		//                         the endpoint is always deleted though, no
-		//                         need to retry
+	req := &models.EndpointBatchDeleteRequest{ContainerID: args.ContainerID}
```

Comment:
Let's not use the new API in v1.14 plugins. The goal of backporting the batch delete API is to be forward-compatible with v1.15. But if v1.14 would use it too, we'd introduce a downgrade hazard to v1.13 (and then we'd have to backport the API to v1.13 too) :sweat_smile:

Reply:
I think EndpointBatchDeleteRequest with containerID is backward and forward compatible, we actually need use it in https://github.com/cilium/cilium/pull/31456 backport.

Reply:
I'm not sure I follow. Let's assume the following scenario:

- User deploys v1.14 with this patch https://github.com/cilium/cilium/pull/31456 and https://github.com/cilium/cilium/pull/27351 applied. That means the CNI plugin will write batch deletion requests into the deletion queue.
- Cilium Agent becomes unavailable for some reason, CNI plugins starts writing batch deletion requests into the deletion queue
- User needs to downgrade to v1.13. That means cilium-agent 1.13 comes up and will observe a deletion queue with batch deletion requests. It will not be able to process them, so the endpoints that are not deleted.

---

That's why we want to backport the _API_ to v1.14, but we do want v1.14 to actually use it. In other words, v1.14 needs to be forward-compatible (by being able to process batch deletes for deletion requests issued by CNI plugin 1.15) and v1.15 needs to be backward-compatible by always using batch deletion (to ensure we delete restored endpoints which do not yet have a cni-attachment-id)

This is why I think we want to do the following:
- Backport https://github.com/cilium/cilium/pull/31456 to v1.15 for the backwards-compatibility, this was already done
- Only backport the server-side of https://github.com/cilium/cilium/pull/27351 to v1.14 for forward-compatibility: cilium-agent 1.14 can understand requests issued by a v1.15 plugin - but it must not issue those requests itself, as otherwise v1.13 would not understand them

Reply:
Thanks for your explanation, it missed the below part. It's making sense to me now.

>  v1.14 needs to be forward-compatible (by being able to process batch deletes for deletion requests issued by CNI plugin 1.15)